Perl_clear_defarray: faster array creation via new macro+function

richardleach opened the pull request:
It's moderately common in core to see the SV* array for an AV created in the
following sort of ways:

```
        AV *av = newAV();
        av_extend(av, key);
```
or
```
        AV * av = newAV();
        av_push(av, some_sv);
```

or
```
        AV * av = newAV();
        av_store(av, 0, some_sv);
```

In each case, array creation ultimately occurs via `av_extend()` and involves
steps that we can reason are unnecessary:
* Check if it's a tied array - we know it isn't
* Call `av_extend_guts()`
* Check` if (key < -1)` - we know it isn't (or can make sure it isn't)
* Check `if (key > *maxp)` - we know it is, because there's no array yet
* Check `if (av && *allocp != *arrayp)` - i.e. is there a shifted array.
                         - we know there isn't, because there's no array yet
* Check `if (*allocp)` - if an unshifted array exists
                         - we know it doesn't , because there's no array yet
* Check `if key < 3`, if so set` key == 3`

Some parts of core already work around this (although sometimes because they
really want less than the minimum key size of 3) by calling `Newx/Newxz` directly.
Examples are:
* `av_make` does it obviously: https://github.com/Perl/perl5/blob/blead/av.c#L421
* https://github.com/Perl/perl5/blob/blead/pad.c#L242 and
https://github.com/Perl/perl5/blob/blead/pad.c#L2596 are variations.
* `new_stackinfo` sort of does it: https://github.com/Perl/perl5/blob/blead/scope.c#L76

This PR does two things:
* Introduces a function and some macros around newAV() + SV* array alloc.
* Modifies `Perl_clear_defarray()` to show an example of usage.

Names and implementation are just best-efforts, happy to change upon review!

richardleach committed:
Add Perl_av_new_alloc() function and newAV_alloc_x/z() macros

richardleach committed:
Perl_clear_defarray: faster array creation via new macro+function

richardleach commented:
Notes:
1. If accepted, I'll submit separate PRs for each function where it seems that unnecessary CPU cycles can be avoided.
2. https://github.com/Perl/perl5/pull/18696 could be altered to use the new macros/functions also. I've left it is for now, since that PR was submitted first, but am happy to revise.

richardleach commented:
Gah. Just fixing the cat-across-the-keyboard reason for the CI build failure...

hvds commented on av.c:
```diff
@@ -393,6 +393,39 @@ Perl_av_store(pTHX_ AV *av, SSize_t key, SV *val)
 }
 
+/*
+=for apidoc av_new_alloc
```
As well as the implementation of `zeroflag`, this has some other differences to the newAV+av_extend equivalent that I think should either be corrected or documented. I notice at least:
- lower bound check (size < 0)
- upper bound check (memwrap)
- defaulting to minimum (max(size, 4))

My first guess is that the lack of default minimum should be documented, but the bounds checks should be added.

richardleach replied:
Yes, I had intended that people using this are certain that they will be calling with a known, non-zero size. I will improve the documentation and add an assert for debugging builds.

I thought that Newx - i.e. safesysmalloc - checks for memwrap?

hvds replied:
> Yes, I had intended that people using this are certain that they will be calling with a known, non-zero size. I will improve the documentation and add an assert for debugging builds.

Ok.

> I thought that Newx - i.e. safesysmalloc - checks for memwrap?

Ah yes, you're quite right - so the `MEM_WRAP_CHECK_s` in `av_extend_guts` is actually a duplicate, presumably only for the purpose of giving a more specific error message. Maybe that could be combined into a `Newx_s` macro to avoid the duplication.

richardleach replied:
Huh, _safesysmalloc_ does not check for memwrap. I think I intended that people using this function are certain that they are using a non-zero, positive integer that is small enough to never wrap! But I'd initially envisaged the usage being for small, hardcoded sizes, which is obviously not the case for the pp_hot.c function changed in this commit.

I'll add a check for memwrap.

richardleach replied:
D'oh, I must have looked more closely the first time around! Let me look again.

richardleach replied:
Yes, I misremembered where the memwrap check was. (i.e. The `Newx` macros, not the `safesysmalloc` function.)

So I propose to not add an extra check into this PR and we open a separate issue for the `Newx_s` macro you suggest. Sound okay?

hvds replied:
> So I propose to not add an extra check into this PR and we open a separate issue for the `Newx_s` macro you suggest. Sound okay?

Yes.

hvds commented on av.h:
```diff
@@ -108,6 +108,26 @@ Perl equivalent: C<my @array;>.
+    av_extend(av, key);
+
+Note that these the actual size requested is allocated. This is unlike
+av_extend(), which enforces a minimum size of 3, and always does a +1
+to the requested/minimum size when allocating.
```
It adds 1 to the _key_, but that's not the "requested/minimum size" but the index of the maximum element. So the minimum key is 3, but minimum size is 4 (covering indices 0-3).

richardleach replied:
Yes, good point. I'll try to improve the wording.

hvds commented on av.h:
```diff
@@ -108,6 +108,26 @@ Perl equivalent: C<my @array;>.
+    AV *av = newAV();
+    av_extend(av, key);
+
+Note that these the actual size requested is allocated. This is unlike
```
s/that these the/that the/

richardleach replied:
Will fix.

hvds commented on pp_hot.c:
```diff
@@ -4975,8 +4975,8 @@ Perl_clear_defarray(pTHX_ AV* av, bool abandon)
-        AV *newav = newAV();
-        av_extend(newav, fill);
+        /* fill can be -1. The ternary gives consistency with av_extend() */
+        AV *newav = newAV_alloc_xz(fill < 3 ? 4 : fill);
```
This looks wrong, because `newAV_alloc_xz()` takes a size rather than a key - it should be `fill < 3 ? 4 : fill + 1`. For cleanliness, it would probably be better instead to set `const SSize_t size = AvFILLp(av) + 1; ... newAV_alloc_xz(size < 4 ? 4 : size)`

Note also that the original `av_extend()` call zero-filled the array only because the `AvREIFY_only()` is applied late; a small optimization therefore would be to use `newAV_alloc_x()` instead. (But since that's a change of behaviour, it's probably worth calling out with a separate commit.)

richardleach replied:
Gah, you're right. Managed to get confused by key vs size myself.

The point about `AvREIFY_only()` is a good one. I'll double check: might well have left it out since it was a change, but half recollect trying to just call `newAV_alloc_x()` and being surprised when that broke things.

richardleach replied:
I was thinking about some other refactoring breaking with `newAV_alloc_x()`. I've added the change to use that as a separate commit (to remain unsquashed at the end).

richardleach committed:
Perl_clear_defarray - AvREIFY_only() applied, array does not actually need Zeroing

richardleach commented:
@hvds - Have I addressed your review points sufficiently, or do you recommend any more changes?

As far as eventual squashing goes, I was thinking of ending up with three commits:
* One to introduce the new macros/function
* One to change `Perl_clear_defarray` to use `newAV_alloc_xz` (i.e. no change in Zeroing behaviour)
* One to change `newAV_alloc_xz` to `newAV_alloc_x` (the change to no array Zeroing by itself)

hvds commented:
> @hvds - Have I addressed your review points sufficiently, or do you recommend any more changes?

You've addressed my points excellently; however:
- the assert in Perl_av_new_alloc should be `assert(size >= 1)`
- the `/* fill can be -1. ... */` comment in Perl_clear_defarray no longer makes sense: I suggest removing that fragment, and just leaving `The ternary gives consistency ...`.

> As far as eventual squashing goes, I was thinking of ending up with three commits:
> 
>     * One to introduce the new macros/function
> 
>     * One to change `Perl_clear_defarray` to use `newAV_alloc_xz` (i.e. no change in Zeroing behaviour)
> 
>     * One to change `newAV_alloc_xz` to `newAV_alloc_x` (the change to no array Zeroing by itself)

Yes, that sounds great.

richardleach commented:
@hvds - thanks for the additional feedback. I've made changes and hopefully squashed correctly.

hvds commented:
@richardleach I don't think the commit rearrangement worked quite right. I've tried to fix it to what I think you intended - please check branch [hv/clear_defarray](https://github.com/Perl/perl5/tree/hv/clear_defarray) to see if you agree.

* moved the assert fix to the commit that introduced it
* ensured that the last commit only changes *_xz() to *_x()
* (cosmetic/doc) modified the name of the macro parameters in embed.fnc from `key` to `size`
* also rebased against latest blead

richardleach commented:
@hvds - thanks for paying such close attention to the details. Yes, your branch does look better than my squashing.

hvds commented:
Super, now merged as 0b1c19ab1c, d2a9e96003 and 97b09acf9d.

iabyn commented:
On Wed, May 26, 2021 at 10:10:27AM -0700, Hugo van der Sanden wrote:
> Super, now merged as 0b1c19ab1c, d2a9e96003 and 97b09acf9d.

I find the docs for the new av_new_alloc() function to be very confusing.
Specifically:

    Note that av_index() takes the desired AvMAX as its key parameter, but
    av_new_alloc() instead takes the desired size (so AvMAX + 1). This
    size must be at least 1.

It's suddenly going off on a tangent about two functions/macros not
mentioned before, i.e. av_index() and AvMAX(). I guess it's trying to make
a point about max index != size, but it is (alas) not very clear.

How about just saying something to the effect that the size parameter
causes AvARRAY to be pre-allocated to hold at least elements 0..(size-1) ?


-- 
If life gives you lemons, you'll probably develop a citric acid allergy.

richardleach commented:
> How about just saying something to the effect that the size parameter causes AvARRAY to be pre-allocated to hold at least elements 0..(size-1) ?

Sure, how about this, to go above the sentence about the zeroflag parameter?
`The size parameter is used to pre-allocate a SV* array large enough to
hold elements 0..(size-1).`

iabyn commented:
How about

The size parameter is used to pre-allocate a SV* array large enough to hold at least elements 0..(size-1). size must be at least 1.

We shouldn't guarantee that it won't allocate more than the requested number of elements, for efficiency purposes.

richardleach commented:
> The size parameter is used to pre-allocate a SV* array large enough to hold at least elements 0..(size-1). size must be at least 1.

Ok. #18959 raised to apply that wording. Thanks for the feedback.
